gpcontrib/arenadata_toolkit/src/track_files.c

```diff
@@ -0,0 +1,1447 @@
+	 */
+	ScanKeyInit(&skey[1],
+				Anum_pg_db_role_setting_setrole,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(InvalidOid));
+	scan = systable_beginscan(rel, DbRoleSettingDatidRolidIndexId, true, NULL, 2, skey);
+
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		bool		isnull;
+		Datum		str_datum;
+
+		str_datum = heap_getattr(tuple, Anum_pg_db_role_setting_setconfig,
+								 RelationGetDescr(rel), &isnull);
+		if (!isnull)
+		{
+			ArrayType  *array;
+			Datum	   *elems;
+			bool	   *nulls;
+			int			nelems;
+
+			array = DatumGetArrayTypeP(str_datum);
+			deconstruct_array(array, TEXTOID, -1, false, 'i',
+							  &elems, &nulls, &nelems);
+			for (int i = 0; i < nelems; i++)
+			{
+				if (nulls[i])
+					continue;
+				char	   *str = TextDatumGetCString(elems[i]);
+
+				if (strncmp(str, "arenadata_toolkit.tracking_schemas=", 35) == 0)
+					current_schemas = pstrdup(str + 35);
+				else if (strncmp(str, "arenadata_toolkit.tracking_relstorages=", 39) == 0)
+					current_relstorages = pstrdup(str + 39);
+				else if (strncmp(str, "arenadata_toolkit.tracking_relkinds=", 36) == 0)
+					current_relkinds = pstrdup(str + 36);
+				pfree(str);
+			}
+		}
+	}
+	systable_endscan(scan);
+	heap_close(rel, RowExclusiveLock);
+
+	if (current_schemas)
+		schema_names = split_string_to_list(current_schemas);
+	else
+		schema_names = split_string_to_list(tracked_schemas);
+	if (current_relstorages)
+		tf_get_global_state.relstorages = list_to_bits(current_relstorages);
+	else
+		tf_get_global_state.relstorages = list_to_bits(tracked_rel_storages);
+	if (current_relkinds)
+		tf_get_global_state.relkinds = list_to_bits(current_relkinds);
+	else
+		tf_get_global_state.relkinds = list_to_bits(tracked_rel_kinds);
+
+	foreach(lc, schema_names)
+	{
+		Oid			nspOid;
+		char	   *name = (char *) lfirst(lc);
+
+		nspOid = GetSysCacheOid1(NAMESPACENAME, CStringGetDatum(name));
+
+		if (!OidIsValid(nspOid))
+		{
+			elog(DEBUG1, "[tracking_get_track] schema \"%s\" does not exist", name);
+			continue;
+		}
+
+		tf_get_global_state.schema_oids = lappend_oid(tf_get_global_state.schema_oids, nspOid);
+	}
+
+	if (schema_names)
+		pfree(schema_names);
+}
+
+
+static bool
+schema_is_tracked(Oid schema)
+{
+	ListCell   *lc;
+
+	if (tf_get_global_state.schema_oids == NIL)
+		return false;
+
+	foreach(lc, tf_get_global_state.schema_oids)
+	{
+		Oid			tracked_schema = lfirst_oid(lc);
+
+		if (tracked_schema == schema)
+			return true;
+	}
+
+	return false;
+}
+
+static bool
+kind_is_tracked(char type, uint64 allowed_kinds)
+{
+	return (allowed_kinds & (1ULL << (type - 'A'))) != 0;
+}
+
+/*
+ * Main function for relation size track acquisition.
+ */
+Datum
+tracking_get_track(PG_FUNCTION_ARGS)
+{
+	FuncCallContext *funcctx;
+	tf_main_func_state_t *state;
+	HeapTuple	result;
+	Datum		datums[GET_TRACK_TUPDESC_LEN];
+	bool		nulls[GET_TRACK_TUPDESC_LEN] = {0};
+	uint32		version = PG_GETARG_INT64(0);
```

Comment:
`uint32` vs `INT64`?

Reply:
**postgresql does not have unsigned int types**. And the intended usage assumes using get_version_function as an argument, which wraps uint32 into pg int64

Reply:
> postgresql does not have unsigned int types

https://github.com/petere/pguint

Reply:
Is it ok to introcude new types in our extension?